### ISSUE TITLE
W1b: `RandomGaussianIllumination` returns an all-NaN image for a `sigma` its own constructor admits (0/0 in `filters.kernels.gaussian`)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4589]
### Describe the bug

`RandomGaussianIllumination` accepts `sigma=0.0` — its own constructor check is
`KORNIA_CHECK(all(0 <= el <= 1 for el in sigma), "sigma of gaussian value must be between 0 and 1.")`
(`kornia/augmentation/_2d/intensity/gaussian_illumination.py:177-180`) — and then returns an image of
**all NaN**, with no warning.

The sibling Gaussian path guards against exactly this: `gaussian_blur2d` and `RandomGaussianBlur` reject
a zero sigma with `sigma must be positive, got tensor([[0., 0.]])`. Only the illumination generator admits it.

### Root cause

`kornia/filters/kernels.py:119-120`:

```python
gauss = torch.exp(-x.pow(2.0) / (2 * sigma.pow(2.0)))
return gauss / gauss.sum(-1, keepdim=True)
```

At `sigma == 0` every exponent is `-inf` (or `0/0` at the mean), so `gauss.sum(-1)` is `0` and the
normalisation is `0 / 0`.

It is not only the exact point. `GaussianIlluminationGenerator.forward`
(`kornia/augmentation/random_generator/_2d/gaussian_illumination.py:110-125`) scales the drawn sigma by
the axis length (`sigma_x = width * sigma`) and, for an **even**-length axis, `gaussian` adds the `+0.5`
half-pixel offset — so no sample sits at the mean and the whole kernel underflows for a small non-zero
sigma as well:

```
gaussian(window=4, sigma=0.02) -> [nan, nan, nan, nan]
gaussian(window=5, sigma=0.02) -> [0.0, 0.0, 1.0, 0.0, 0.0]
```

### Reproduction steps

```python
import torch
from kornia.augmentation import RandomGaussianIllumination

torch.manual_seed(0)
aug = RandomGaussianIllumination(p=1.0, sigma=0.0)   # admitted by the class's own KORNIA_CHECK
print(aug(torch.rand(1, 3, 4, 4))[0, 0])
```

```
tensor([[nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan]])
```

All four dtypes NaN (`float16`, `bfloat16`, `float32`, `float64`), every image size tried
(`1x1, 2x2, 3x3, 4x4, 8x8, 32x32`), 100% of pixels.

NaN hits out of 30 seeds, `sigma=(s, s)` on an `n x n` image, default `center`, CPU:

| n \ s | 0.0 | 0.001 | 0.005 | 0.01 | 0.02 | 0.05 | 0.1 |
|---|---|---|---|---|---|---|---|
| 1 | 30 | 26 | 26 | 26 | 26 | 26 | 0 |
| 2 | 30 | 30 | 30 | 30 | 0 | 0 | 0 |
| 3 | 30 | 5 | 5 | 5 | 5 | 0 | 0 |
| 4 | 30 | 30 | 30 | 0 | 0 | 0 | 0 |
| 8 | 30 | 30 | 0 | 0 | 0 | 0 | 0 |
| 32 | 30 | 30 | 0 | 0 | 0 | 0 | 0 |

The even rows lose the on-grid sample to the half-pixel offset, which is why `n=2` and `n=4` NaN at a
sigma where `n=3` does not.

The class default `sigma=(0.2, 1.0)` is clean: 200 seeds x sizes 1..4, zero NaN. `RandomLinearIllumination`
and `RandomLinearCornerIllumination` have no Gaussian kernel and are unaffected.

### Expected behavior

Either reject a sigma that cannot produce a normalizable kernel, the way `gaussian_blur2d` already does,
or make `gaussian` fall back to a unit impulse at the mean. A constructor that admits a value and then
silently emits NaN is the worst of the three.

The narrow fix is a positivity guard in `GaussianIlluminationGenerator` mirroring `gaussian_blur2d`'s;
the broader one is to make `kornia.filters.kernels.gaussian` refuse a non-positive sigma, since its
`0 / 0` is reachable from any caller that does not pre-check.

### Environment

kornia at `8aca2838f` (PR #4566 head), torch 2.14.0, Python 3.11.14, macOS arm64, CPU.
Found while reviewing #4566; the batch 6c convention block for this class currently states that the
output "stays inside `[0, 1]` even when the input does not", which this falsifies.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
Sharpening the mechanism, because the "even-length axis" framing above is narrower than what happens and
would send a fix at the wrong variable.

Evenness decides whether a sample sits **at** the mean; it does not decide whether the kernel underflows.
The generator scales the drawn sigma by the axis length (`sigma_x = width * sigma`), so with
`center=(0.5, 0.5)` the nearest tap on an even axis is half a pixel off the mean and the exponent is
`-(0.5 ** 2) / (2 * (n * sigma) ** 2)`. It is `n * sigma` — the kernel's **absolute** width — that
crosses `exp`'s underflow:

```python
import torch, kornia.augmentation as K

for n, sg in ((2, 0.005), (4, 0.005), (6, 0.005), (8, 0.005), (16, 0.005), (64, 0.005),
              (64, 0.0005), (64, 0.00055), (8, 0.002)):
    torch.manual_seed(0)
    out = K.RandomGaussianIllumination(p=1.0, sigma=(sg, sg), center=(0.5, 0.5))(
        torch.full((1, 3, n, n), 0.5))
    print(f"n={n:3d} sigma={sg:<8} n*sigma={n * sg:.4f} -> {'NaN' if out.isnan().any() else 'finite'}")
```

```
n=  2 sigma=0.005    n*sigma=0.0100 -> NaN
n=  4 sigma=0.005    n*sigma=0.0200 -> NaN
n=  6 sigma=0.005    n*sigma=0.0300 -> NaN
n=  8 sigma=0.005    n*sigma=0.0400 -> finite
n= 16 sigma=0.005    n*sigma=0.0800 -> finite
n= 64 sigma=0.005    n*sigma=0.3200 -> finite
n= 64 sigma=0.0005   n*sigma=0.0320 -> NaN
n= 64 sigma=0.00055  n*sigma=0.0352 -> finite
n=  8 sigma=0.002    n*sigma=0.0160 -> NaN
```

So a `64 x 64` image — even — is finite at the sigma that kills `4 x 4`, and dies at a proportionally
smaller one. The cut sits at `n * sigma ~= 0.034`.

Two more details a fix needs:

- **The threshold follows the parameter dtype, not the image dtype.** The gradient is built in the
  generator's dtype, which is `float32` for a `float16`, `bfloat16`, `float32` *or* `float64` image
  (`_params["gradient"].dtype` is `torch.float32` in all four). After
  `aug.set_rng_device_and_dtype(device, torch.float64)`, the `4 x 4` / `sigma=0.005` case is **finite**
  and `0.003` is not.
- **An odd axis is not unconditionally safe, but it is safe from 5 up.** With the default
  `center=(0.1, 0.9)` the rounded centre can leave the grid only on a 1- or 3-pixel axis: over 30 seeds at
  `sigma=0.005`, `n=1` is NaN on 23, `n=3` on 14, and `n=5`, `7`, `9` on none.

A guard on `gauss.sum()` (or a `KORNIA_CHECK` that the scaled sigma is representable) covers all of it;
a guard keyed on `window_size % 2` would not.

Measured on `90650596` (main), torch 2.14.0, Python 3.11, macOS arm64, CPU.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
